addons/stock/models/stock_move.py

```diff
@@ -664,22 +672,23 @@ def write(self, vals):
             self._set_date_deadline(vals.get('date_deadline'))
         if 'move_orig_ids' in vals:
             move_to_recompute_state |= self.filtered(lambda m: m.state not in ['draft', 'cancel', 'done'])
-        if 'location_dest_id' in vals:
-            move_to_check_dest_location = self.filtered(lambda m: m.location_dest_id.id != vals.get('location_dest_id'))
+        if 'location_id' in vals:
+            move_to_check_location = self.filtered(lambda m: m.location_id.id != vals.get('location_id'))
         if 'picking_id' in vals and 'group_id' not in vals:
             picking = self.env['stock.picking'].browse(vals['picking_id'])
             if picking.group_id:
                 vals['group_id'] = picking.group_id.id
         res = super(StockMove, self).write(vals)
         if move_to_recompute_state:
             move_to_recompute_state._recompute_state()
-        if move_to_check_dest_location:
-            for ml in move_to_check_dest_location.move_line_ids:
-                parent_path = [int(id) for id in ml.location_dest_id.parent_path.split('/')[:-1]]
-                if ml.move_id.location_dest_id.id in parent_path:
-                    continue
-                loc_dest = ml.move_id.location_dest_id._get_putaway_strategy(ml.product_id, ml.quantity_product_uom)
-                ml.location_dest_id = loc_dest
+        if move_to_check_location and self.env.context.get('do_not_unreserve') is None:
```

Comment:
I'm not too fond of that check. It seems weird that it does the same thing if `do_not_unreserve` is either `True` or `False`, but not if the key isn't given. Imho this may require another key, rather than creating a hacky side-effect for an existing key.